cmd/earthly/main.go

```diff
@@ -966,6 +1008,66 @@ func (app *earthlyApp) before(context *cli.Context) error {
 	return nil
 }
 
+func (app *earthlyApp) getBuildkitAndDebuggerAddressesOriginal() (string, string, error) {
+	dbAddr := fmt.Sprintf("tcp://127.0.0.1:%v", app.debuggerPort)
+	return app.buildkitHost, dbAddr, nil
+}
+
+func (app *earthlyApp) getBuildkitAndDebuggerAddressesForTCP(context *cli.Context) (string, string, error) {
+	if !context.IsSet("buildkit-host") {
+		if app.cfg.Global.BuildkitHost != "" {
+			app.buildkitHost = app.cfg.Global.BuildkitHost
+		} else {
+			app.buildkitHost = buildkitd.TCPAddress
+		}
+	}
+
+	bkURL, err := parseAndvalidateURL(app.buildkitHost)
+	if err != nil {
+		return "", "", err
+	}
+
+	if !context.IsSet("debugger-host") {
+		if app.cfg.Global.DebuggerHost != "" {
+			app.debuggerHost = app.cfg.Global.DebuggerHost
+		} else {
+			app.debuggerHost = fmt.Sprintf("tcp://%s:%v", bkURL.Hostname(), config.DefaultDebuggerPort)
+		}
+	}
+
+	dbURL, err := parseAndvalidateURL(app.debuggerHost)
+	if err != nil {
+		return "", "", err
+	}
+
+	if bkURL.Hostname() != dbURL.Hostname() {
+		app.console.Warnf("Buildkit and Debugger URLs are pointed at different hosts (%s vs. %s)", bkURL.Hostname(), dbURL.Hostname())
+	}
+
+	if bkURL.Hostname() == dbURL.Hostname() && bkURL.Port() == dbURL.Port() {
+		return "", "", errors.New("Debugger and Buildkit ports are the same")
+	}
+
+	return app.buildkitHost, app.debuggerHost, nil
+}
+
+func parseAndvalidateURL(addr string) (*url.URL, error) {
+	parsed, err := url.Parse(addr)
+	if err != nil {
+		return nil, errors.Errorf("%s is not a valid URL", addr)
+	}
+
+	if parsed.Scheme != "tcp" {
+		return nil, errors.Errorf("%s is not a valid scheme. Only tcp is allowed at this time.", parsed.Scheme)
+	}
+
+	if parsed.Port() == "" {
+		return nil, errors.Errorf("%s does not contain a port number.", addr)
```

Comment:
This feels like it shouldn't be an error. URLs without a port are still valid? Or maybe this is because it's `tcp://` and you can't infer a port unless it's `http://` or `https://`. Yeah, I think I gave myself the explanation :D